File: ios/Tehilim/Core/Services/HebrewNumerals.swift
import Foundation

/// Conversion entre entiers (1...400) et notation hébraïque par gematria.
/// Respecte les conventions standards : 15 = טו, 16 = טז (et non יה / יו).
enum HebrewNumerals {

    private static let mapping: [(value: Int, letter: String)] = [
        (400, "ת"), (300, "ש"), (200, "ר"), (100, "ק"),
        (90,  "צ"), (80,  "פ"), (70,  "ע"), (60,  "ס"),
        (50,  "נ"), (40,  "מ"), (30,  "ל"), (20,  "כ"),
        (10,  "י"),
        (9,   "ט"), (8,   "ח"), (7,   "ז"), (6,   "ו"),
        (5,   "ה"), (4,   "ד"), (3,   "ג"), (2,   "ב"), (1, "א")
    ]

    /// Convertit un entier en lettres hébraïques (sans gershayim).
    static func toHebrew(_ n: Int) -> String {
        guard n > 0 else { return "" }
        var remainder = n
        var result = ""

        // Cas spéciaux 15 et 16 — gérer la dizaine et garder l'unité spéciale.
        // On traite les centaines normalement, puis le reste avec règle 15/16.
        if remainder >= 100 {
            for (val, letter) in mapping where val >= 100 {
                while remainder >= val {
                    result += letter
                    remainder -= val
                }
            }
        }
        if remainder == 15 { return result + "טו" }
        if remainder == 16 { return result + "טז" }

        for (val, letter) in mapping where val < 100 {
            while remainder >= val {
                result += letter
                remainder -= val
            }
        }
        return result
    }

    /// Convertit une chaîne hébraïque en entier (best-effort).
    /// - Ignore les caractères non-hébreu (gershayim, espaces, ponctuation).
    static func toInt(_ s: String) -> Int? {
        let cleaned = s.unicodeScalars.filter { isHebrewLetter($0) }.map(Character.init).map(String.init).joined()
        guard !cleaned.isEmpty else { return nil }

        let values: [Character: Int] = [
            "א": 1, "ב": 2, "ג": 3, "ד": 4, "ה": 5,
            "ו": 6, "ז": 7, "ח": 8, "ט": 9, "י": 10,
            "כ": 20, "ך": 20, "ל": 30, "מ": 40, "ם": 40,
            "נ": 50, "ן": 50, "ס": 60, "ע": 70, "פ": 80, "ף": 80,
            "צ": 90, "ץ": 90, "ק": 100, "ר": 200, "ש": 300, "ת": 400
        ]

        var total = 0
        for ch in cleaned {
            guard let v = values[ch] else { return nil }
            total += v
        }
        return total > 0 ? total : nil
    }

    private static func isHebrewLetter(_ scalar: Unicode.Scalar) -> Bool {
        // Plage des lettres hébraïques + finales
        return (0x05D0...0x05EA).contains(scalar.value)
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/HebrewNumerals.kt
package com.david.tehilim.core.service

/**
 * Conversion entre entiers (1..400) et notation hébraïque par gematria.
 * Respecte les conventions standards : 15 = טו, 16 = טז (et non יה / יו).
 *
 * Mirror direct de `HebrewNumerals.swift` iOS.
 */
object HebrewNumerals {

    private val mapping: List<Pair<Int, String>> = listOf(
        400 to "ת", 300 to "ש", 200 to "ר", 100 to "ק",
        90 to "צ", 80 to "פ", 70 to "ע", 60 to "ס",
        50 to "נ", 40 to "מ", 30 to "ל", 20 to "כ",
        10 to "י",
        9 to "ט", 8 to "ח", 7 to "ז", 6 to "ו",
        5 to "ה", 4 to "ד", 3 to "ג", 2 to "ב", 1 to "א"
    )

    /** Convertit un entier en lettres hébraïques (sans gershayim). */
    fun toHebrew(n: Int): String {
        if (n <= 0) return ""
        var remainder = n
        val sb = StringBuilder()

        // Centaines
        if (remainder >= 100) {
            for ((value, letter) in mapping.filter { it.first >= 100 }) {
                while (remainder >= value) {
                    sb.append(letter)
                    remainder -= value
                }
            }
        }
        // Cas spéciaux 15 / 16
        if (remainder == 15) return sb.append("טו").toString()
        if (remainder == 16) return sb.append("טז").toString()

        for ((value, letter) in mapping.filter { it.first < 100 }) {
            while (remainder >= value) {
                sb.append(letter)
                remainder -= value
            }
        }
        return sb.toString()
    }

    /**
     * Convertit une chaîne hébraïque en entier (best-effort).
     * Ignore les caractères non-hébreu (gershayim, espaces, ponctuation).
     */
    fun toInt(s: String): Int? {
        val cleaned = s.filter { it.code in 0x05D0..0x05EA }
        if (cleaned.isEmpty()) return null

        val values: Map<Char, Int> = mapOf(
            'א' to 1, 'ב' to 2, 'ג' to 3, 'ד' to 4, 'ה' to 5,
            'ו' to 6, 'ז' to 7, 'ח' to 8, 'ט' to 9, 'י' to 10,
            'כ' to 20, 'ך' to 20, 'ל' to 30, 'מ' to 40, 'ם' to 40,
            'נ' to 50, 'ן' to 50, 'ס' to 60, 'ע' to 70,
            'פ' to 80, 'ף' to 80, 'צ' to 90, 'ץ' to 90,
            'ק' to 100, 'ר' to 200, 'ש' to 300, 'ת' to 400
        )

        var total = 0
        for (ch in cleaned) {
            val v = values[ch] ?: return null
            total += v
        }
        return if (total > 0) total else null
    }
}
